riot.py

```diff
@@ -0,0 +1,480 @@
+"""
+TODO
+- json report for verify_all_tests
+- timing metadata (how long it takes to setup and run a suite, case, etc)
+- format/naming convention for cases
+- parallelize virtualenv creation - there's no reason to not to
+- intermediate venv building eg. if cases share deps X, Y, Z, generate an env
+  for these and copy it
+- does specifying encoding="utf-8" affect the locale of the test cases?
+- reporting to datadog (stats, traces)
+- typing
+- clean up into classes w/ methods?
+- better log formatter / output
+- better handling of CmdFailure in run_suites
+- add command to run pytest (or any command) in a venv
+  eg: python riot.py --suite tracer --python 3.7 --command pytest -k test_tracer_fork
+"""
+import argparse
+import itertools
+import logging
+import os
+import re
+import shutil
+import subprocess
+import sys
+import typing as t
+
+
+logger = logging.getLogger(__name__)
+
+
+SHELL = "/bin/bash"
+ENCODING = "utf-8"
+
+
+class AttrDict(dict):
+    def __init__(self, *args, **kwargs):
+        super(AttrDict, self).__init__(*args, **kwargs)
+        self.__dict__ = self
+
+
+class CmdFailure(Exception):
+    def __init__(self, msg, completed_proc):
+        self.msg = msg
+        self.proc = completed_proc
+        self.code = completed_proc.returncode
+        super().__init__(self, msg)
+
+
+# It's easier to read `Suite`, `Case` rather than AttrDict or dict.
+CaseInstance = Case = Suite = AttrDict
+
+global_deps = [
+    "mock",
+    "opentracing",
+    "pytest<4",
+    "pytest-benchmark",
+    "pytest-django",
+]
+
+global_env = [("PYTEST_ADDOPTS", "--color=yes")]
+
+all_suites = [
+    Suite(
+        name="tracer",
+        command="pytest tests/test_tracer.py",
+        env=dict(),
+        cases=[
+            Case(pys=[2.7, 3.5, 3.6, 3.7, 3.8,], pkgs=[("msgpack", [None, "==0.5.0", ">=0.5,<0.6", ">=0.6.0,<1.0"])],),
```

Comment:
would it make sense to define "case defaults"?

e.g.

```python
Suite(
    case_defaults=CaseDefaults(pys=[2.7, 3.5, 3.6, 3.7, 3.8]),
    cases=[
    ]
)
```

Then we only need to supply `pys` when it is different from the default ?

Reply:
Hmm yeah I'm on the fence about introducing too much indirection. IMO it leads to additional overhead when trying to read the configuration. If we notice that it becomes too unmaintainable when porting over the rest of the configurations from tox, then I think we can reconsider.